### PR TITLE
Pattern Matching / Catamorphisms

### PR DESCRIPTION
**Pull request by ostera**

### CODE REVIEW

File: src/cond.js [hunk 1]
```diff
@@ -27,6 +27,30 @@ const reducer: Reducer = (a, cond) => a || run_cond(cond) || a
...
+
+const cata = match({
+  Inc: x => x+1,
+  Dec: x => x-1,
+  Reset: 0
+})
+
+console.log(cata(IncAction.of(10)))
+console.log(cata(DecAction.of(10)))
+console.log(cata(WhatAction.of(100)))
```
Comment (ostera) [inline src/cond.js]: screen shot 2017-02-04 at 12 40 25 am

File: src/arrow.js [hunk 2]
```diff
@@ -53,6 +53,7 @@ const id = x => x
 // Lifts a function into an arr
 // arrrow :: (b -> c) -> arr b c
 const Arrow = (f: Function): ArrowT  => {
+  f[Symbol.for('TypeName')] = 'Arrow'
```
Comment (xaviervia) [inline src/arrow.js]: Let's use `@@type` here

**Comment by xaviervia** [comment]
Pending: 

- [ ] Add `is` to some types to create predicates for refined types (https://medium.com/@gcanti/refinements-with-flow-9c7eeae8478b#.g4jijkgw4)
  Can be something like:
  ```javascript
  const createSetoid = (is) => (name) => ({
    …
    is
  })
  ```

- [ ] Test nesting `match` calls with refined types.


Raw source: 

```javscript
/*
 * Generic Type Creator. If used with Flow's Inference engine, works smoothly
 * to verify that `of` blows up in time for the wrong types.
 */
const createType = (name: any): any => ({
  of: x => ({
    '@@type': name,
    '@@value': x,
  }),

  is: (x) => bool
})

const createFunctor = (map: Map) => (name: any): any => ({
  of: x => ({
    '@@type': name,
    '@@value': x,
    map: map(x)
  })
})

export {
  createType,
  createFunctor,
}

const Above100: Data<Above100T, number> = createDependantType(x => x > 100)('Above100')

match({
  Left: match({
    Number:
  })
})
```

File: src/type.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+const createType = (name: any): any => ({
+  of: x => ({
+    '@@type': name,
+    '@@value': x,
+  })
+})
+
+const createFunctor = (map: Map) => (name: any): any => ({
+  of: x => ({
+    '@@type': name,
```
Comment (ostera) [inline src/type.js]: Can we use `...createType(name).of(x),` instead of redefining `@@type` and `@@value` ?

**Comment by ostera** [comment]
👍

**Pull request merged**